# RiceG2G  

### General Introduction 
GWAS in rice is difficult to reach single-gene resolution due to modest linkage disequilibrium decay.In order to systematically prioritize causal genes and variants at trait-associated loci, we developed an integrated genomics approach named RiceG2G. RiceG2G is an integrated genomics approach to systematically prioritize causal genes and variants at trait-associated loci. Which combined genetic associations with gene annotation, transcriptomics, functional genomics and intolerant variants, and produced a well-calibrated score. Briefly speaking, for each gene at a given associated locus, we collected its annotated type , expression pattern in vast spatio-temporal tissues, functional studies of homologs in other plants, variant interpretation, linkage mapping signal in each family and variant-signal correlations among families. Depending on the given trait, a summarized G2G score was then calculated for each candidate gene, and the score rank of candidate genes at an associated locus could reflect the likelihood of any certain gene to be the causal one. Moreover, once a strong candidate gene was determined, the most possible causal QTN could be provided based on variant interpretation, owing to nearly complete information of genetic variants from genome assemblies of parental lines.
### Availability
You can download the source code and the Demo data here:
[RiceG2G](https://github.com/xhhuanglab/RiceG2G)

###  Installation
```
git clone https://github.com/xhhuanglab/RiceG2G
```
